cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/SourceControlMetadataRefreshService.java

```diff
@@ -0,0 +1,179 @@
+ * NamespaceSourceControlMetadataRefreshService scheduler instances.
+ */
+public class SourceControlMetadataRefreshService extends AbstractIdleService {
+
+  private ConcurrentMap<NamespaceId, NamespaceSourceControlMetadataRefreshService>
+      refreshSchedulers = new ConcurrentHashMap<>();
+  private final TransactionRunner transactionRunner;
+  private final NamespaceAdmin namespaceAdmin;
+  private final CConfiguration cConf;
+  private final SourceControlOperationRunner sourceControlOperationRunner;
+  private static final Logger LOG = LoggerFactory.getLogger(
+      SourceControlMetadataRefreshService.class);
+  private final FeatureFlagsProvider featureFlagsProvider;
+
+  /**
+   * Constructs a SourceControlMetadataRefreshService instance.
+   */
+  @Inject
+  public SourceControlMetadataRefreshService(TransactionRunner transactionRunner,
+      NamespaceAdmin namespaceAdmin, CConfiguration cConf,
+      SourceControlOperationRunner sourceControlOperationRunner,
+      Store store) {
+    this.transactionRunner = transactionRunner;
+    this.namespaceAdmin = namespaceAdmin;
+    this.cConf = cConf;
+    this.sourceControlOperationRunner = sourceControlOperationRunner;
+    this.featureFlagsProvider = new DefaultFeatureFlagsProvider(cConf);
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    if(!isSourceControlMetadataPeriodicRefreshFlagEnabled()) {
+      return;
+    }
+    LOG.info("Starting SourceControlMetadataRefreshService");
+
+    namespaceAdmin.list().stream()
+        .map(NamespaceMeta::getNamespaceId)
+        .filter(namespaceId -> getRepositoryMeta(namespaceId) != null)
+        .forEach(this::addRefreshService);
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.info("Shutting down SourceControlManagementService");
+
+    for (NamespaceSourceControlMetadataRefreshService service : refreshSchedulers.values()) {
+      service.stopAndWait();
+    }
+  }
+
+  private NamespaceSourceControlMetadataRefreshService createRefreshService(NamespaceId namespace) {
+    return new NamespaceSourceControlMetadataRefreshService(cConf, transactionRunner,
+        sourceControlOperationRunner, namespace, namespaceAdmin);
+  }
+
+  /**
+   * Adds {@link NamespaceSourceControlMetadataRefreshService} for the specified namespace. If not
+   * created before, it starts the service or else, triggers a forced manual refresh
+   *
+   * @param namespaceId The ID of the namespace for which to add the refresh service.
+   */
+  public void addRefreshService(NamespaceId namespaceId) {
+    NamespaceSourceControlMetadataRefreshService service =
+        refreshSchedulers.putIfAbsent(namespaceId, createRefreshService(namespaceId));
+    if (service == null || isSourceControlMetadataPeriodicRefreshFlagEnabled()) {
+      refreshSchedulers.get(namespaceId).start();
+      LOG.debug("Added Scheduled NamespaceSourceControlMetadataRefreshService for "
+          + namespaceId.getNamespace());
+    } else {
```

Comment:
why are we triggering here again ?

Reply:
```
 public void addRefreshService(NamespaceId namespaceId) {
    NamespaceSourceControlMetadataRefreshService service =
        refreshSchedulers.putIfAbsent(namespaceId, createRefreshService(namespaceId));
    if (service == null || isSourceControlMetadataPeriodicRefreshFlagEnabled()) {
      refreshSchedulers.get(namespaceId).start();
    } else {
      refreshSchedulers.get(namespaceId).triggerManualRefresh(true);
    }
  }
  ```
  If we are setting the repo config for the first time, it creates refresh service, calls start function, and performs runOneIteration.
  When we are updating the repo config, it is already present in `refreshSchedulers`. If start() function is called, it will see that it has already started and will *not* call runOneIteration instantly. So we have to trigger it manually so that the changes are reflected in the git pipelines page.